core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -0,0 +1,614 @@
+      scheduleTasks(queuedDrivers,
+        removeFromQueuedDrivers,
+        currentOffers, tasks, currentTime)
+    }
+    tasks.foreach { case (offerId, tasks) =>
+      driver.launchTasks(Collections.singleton(offerId), tasks)
+    }
+    offers
+      .filter(o => !tasks.keySet.contains(o.getId))
+      .foreach(o => driver.declineOffer(o.getId))
+  }
+
+  def getState(): MesosClusterSchedulerState = {
+    def copyBuffer(
+      buffer: ArrayBuffer[MesosDriverDescription]): ArrayBuffer[MesosDriverDescription] = {
+      val newBuffer = new ArrayBuffer[MesosDriverDescription](buffer.size)
+      buffer.copyToBuffer(newBuffer)
+      newBuffer
+    }
+    stateLock.synchronized {
+      new MesosClusterSchedulerState(
+        frameworkId,
+        masterInfo.map(m => s"http://${m.getIp}:${m.getPort}"),
+        copyBuffer(queuedDrivers),
+        launchedDrivers.values.map(_.copy).toList,
+        finishedDrivers.map(_.copy).toList,
+        copyBuffer(pendingRetryDrivers))
+    }
+  }
+
+  def getStatus(submissionId: String): SubmissionStatusResponse = {
+    val s = new SubmissionStatusResponse
+    if (!ready) {
+      s.success = false
+      s.message = "Scheduler is not ready to take requests"
+      return s
+    }
+    s.submissionId = submissionId
+    stateLock.synchronized {
+      if (queuedDrivers.exists(_.submissionId.equals(submissionId))) {
+        s.success = true
+        s.driverState = "Driver is queued for launch"
+      } else if (launchedDrivers.contains(submissionId)) {
+        s.success = true
+        s.driverState = "Driver is running"
+        launchedDrivers(submissionId).taskState.foreach(state => s.message = state.toString)
+      } else if (finishedDrivers.exists(s => s.submission.submissionId.equals(submissionId))) {
+        s.success = true
+        s.driverState = "Driver already finished"
+        finishedDrivers.find(d => d.submission.submissionId.equals(submissionId)).get.taskState
+          .foreach(state => s.message = state.toString)
+      } else if (pendingRetryDrivers.exists(_.submissionId.equals(submissionId))) {
+        val status = pendingRetryDrivers.find(_.submissionId.equals(submissionId))
+          .get.retryState.get.lastFailureStatus
+        s.success = true
+        s.driverState = "Driver failed and retrying"
+        s.message = status.toString
+      } else {
+        s.success = false
+        s.driverState = "Driver not found"
+      }
+    }
+    s
+  }
+
+  override def offerRescinded(driver: SchedulerDriver, offerId: OfferID): Unit = {}
+  override def disconnected(driver: SchedulerDriver): Unit = {}
+  override def reregistered(driver: SchedulerDriver, masterInfo: MasterInfo): Unit = {
+    logInfo(s"Framework re-registered with master ${masterInfo.getId}")
+  }
+  override def slaveLost(driver: SchedulerDriver, slaveId: SlaveID): Unit = {}
+  override def error(driver: SchedulerDriver, error: String): Unit = {}
```

Comment:
This is kind of scary since we're swallowing potential error messages. Perhaps we should `logError` these messages instead.